### PR TITLE
Household type calculations must be made against full household

### PR DESCRIPTION
**Pull request by eanders**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

This updates the core demographics and demographic summary report such that if you run it with a filter that limits the clients included in the report based on client details, the household types are still calculated correctly.

## Type of change
[//]: # 'remove options that are not relevant'
- [x] Bug fix

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb [hunk 1]
```diff
@@ -74,10 +74,16 @@ def household_type_data_for_export(rows)
     end
 
     private def enrollment_ids_in_household_type(key)
+      # Force calculation of Households if necessary
+      households
+
```
Comment (awisema) [inline drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb]: Maybe, move this down into `client_households` to make sure it always gets done when needed?

File: drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb [hunk 2]
```diff
@@ -74,10 +74,16 @@ def household_type_data_for_export(rows)
...
       client_households[key]&.values&.flatten
     end
 
     private def hoh_enrollment_ids_in_household_type(key)
+      # Force calculation of Households if necessary
+      hoh_enrollments
+
```
Comment (awisema) [inline drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb]: Same for `hoh_enrollments`?

File: drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb [hunk 3]
```diff
@@ -89,8 +95,10 @@ def available_household_types
       @hoh_enrollments ||= {}
       @households ||= {}
 
+      # Ignore client-specific filters so we can calculate household type based on who was there, not who is availalble for reporting
```
Comment (awisema) [inline drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb]:
```suggestion
      # Ignore client-specific filters so we can calculate household type based on who was there, not who is available for reporting
```

File: drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb [hunk 4]
```diff
@@ -158,39 +166,52 @@ def available_household_types
...
+          clients[:unaccompanied_youth] = convert_enrollments_to_client_id_enrollment_id_pairs(filter_enrollments_for_report_scope(unaccompanied_youth_households.values.flatten))&.group_by(&:shift)
+          clients[:unknown] = convert_enrollments_to_client_id_enrollment_id_pairs(filter_enrollments_for_report_scope(unknown_households.values.flatten))&.group_by(&:shift)
         end
       end
     end
 
+    # We need all related household members to calculate the household, but once we know the household type,
+    # we only want the enrollments that meet the report filter criteria
+    private def filter_enrollments_for_report_scope(enrollments, hoh_only: false)
+      enrollments&.dup&.select do |en|
```
Comment (awisema) [inline drivers/core_demographics_report/app/models/core_demographics_report/household_type_calculations.rb]: why is `.dup` required? Pretty sure `select` already creates a new array?

**Review by awisema — APPROVED** [review]

**Review by dtgreiner — APPROVED** [review]

**Pull request merged**